camel/functions/openai_function.py

```diff
@@ -13,42 +13,150 @@
-    https://platform.openai.com/docs/guides/gpt/function-calling. If
-    :obj:`description` and :obj:`parameters` are both :obj:`None`, try to use
-    document parser to generate them.
+    https://platform.openai.com/docs/api-reference/chat/create.
+
+    By default, the tool schema will be parsed from the func, or you can
+    provide a user-defined tool schema to override.
 
-    # flake8: noqa :E501
     Args:
-        func (Callable): The function to call.
-        name (str, optional): The name of the function to be called. Must be
-            a-z, A-Z, 0-9, or contain underscores and dashes, with a maximum
-            length of 64. If :obj:`None`, use the name of :obj:`func`. 
+        func (Callable): The function to call.The tool schema is parsed from
+            the signature and docstring by default.
+        openai_tool_schema (Optional[Dict[str, Any]], optional): user-defined
+            openai tool schema to override the default result.
             (default: :obj:`None`)
-        description (str, optional): The description of what the
-            function does. (default: :obj:`None`)
-        parameters (dict, optional): The parameters the
-            functions accepts, described as a JSON Schema object. See the
-            `Function calling guide <https://platform.openai.com/docs/guides/gpt/function-calling>`_
-            for examples, and the `JSON Schema reference <https://json-schema.org/understanding-json-schema/>`_
-            for documentation about the format.
     """
 
-    def __init__(self, func: Callable, name: Optional[str] = None,
-                 description: Optional[str] = None,
-                 parameters: Optional[Dict[str, Any]] = None):
+    def __init__(
+        self,
+        func: Callable,
+        openai_tool_schema: Optional[Dict[str, Any]] = None,
+    ):
         self.func = func
-        self.name = name or func.__name__
+        self.openai_tool_schema = (openai_tool_schema
+                                   or get_openai_tool_schema(func))
+        self.properties = self.openai_tool_schema
+
+    @staticmethod
+    def validate_openai_tool_schema(openai_tool_schema):
+        r"""
+        {
+            "type": "function",
+            "function": {
+                "name": "function_name",
+                "description": "xxx",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "param1": {
+                            "type": "string",
+                            "description": "xxx",
+                        },
+                        "param2": {
+                            "type": "string",
+                            "description": "xxx",
+                        },
+                    },
+                    "required": ["param1", "param1"],
+                },
+            },
+        }
+        """
+        # 1. check the basic format
```

Comment:
`openai` lib provides a typed dict to check function definition. It can be more clear than checking it manually. Please check: `from openai.types.shared_params import FunctionDefinition, FunctionParameters`

`from openai.types.chat.chat_completion_tool_param import ChatCompletionToolParam` This model may also be useful.

Reply:
Thank you for your reminder, that's a good idea. In the future, I will use 'openai.types.beta.threads.run.ToolAssistantToolsFunction' to validate the format